### PR TITLE
Support transactions

### PR DESCRIPTION
**Pull request by functionaldude**

### CODE REVIEW

File: src/main/kotlin/com/moshbit/katerbase/MongoDatabase.kt [hunk 1]
```diff
@@ -227,9 +293,15 @@ open class MongoDatabase(
       }
 
       thread(isDaemon = true, name = "katerbase-watch-${name}-${Random().nextInt().absoluteValue}") {
+        fun watch() = if (session != null) {
+          internalCollection.watch(session, listOf(pipeline))
+        } else {
+          internalCollection.watch(listOf(pipeline))
+        }
```
Comment (vonox7) [inline src/main/kotlin/com/moshbit/katerbase/MongoDatabase.kt:298]: why always use 2 separate function calls in this PR? you can always pass the session as first argument. (here and everywhere)

File: src/test/kotlin/ReplicaSetDatabaseTests.kt [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+import org.junit.jupiter.api.BeforeAll
+import org.junit.jupiter.api.Test
+
+/**
+ * These tests require a replica set as [testDb]. Look for setup instructions in the "local-development" folder.
+ */
+class ReplicaSetDatabaseTests {
+
+  @Test
+  fun transactionTest() = runBlocking {
```
Comment (vonox7) [inline src/test/kotlin/ReplicaSetDatabaseTests.kt]:
```suggestion
  fun transactionSingleCollectionTest() = runBlocking {
```

Please also provide a test with a transaction that spans over multiple collections :)

File: src/main/kotlin/com/moshbit/katerbase/MongoDatabase.kt [hunk 3]
```diff
@@ -70,6 +71,58 @@ open class MongoDatabase(
...
+    // Sessions time out after 30 minutes, but close it now to save resources
+    // https://www.mongodb.com/docs/manual/reference/command/startSession/
+    transactionalDatabase.session.use { session ->
+      runOnIo { session.startTransaction() }
+      try {
+        action(transactionalDatabase)
+        runOnIo { session.commitTransaction() }
+      } catch (throwable: Throwable) {
+        runOnIo {
+          runOnIo { session.abortTransaction() } // Do not commit transaction to DB in case of exception.
```
Comment (vonox7) [inline src/main/kotlin/com/moshbit/katerbase/MongoDatabase.kt]: 2x runOnIo

File: src/main/kotlin/com/moshbit/katerbase/MongoDatabase.kt [hunk 4]
```diff
@@ -70,6 +71,58 @@ open class MongoDatabase(
 
   inline fun <reified T : MongoMainEntry> getSuspendingCollection() = getSuspendingCollection(entryClass = T::class)
 
+  /**
+   * Use the [TransactionalDatabase]s to modify the DB state in a transaction.
+   * See https://www.mongodb.com/docs/manual/core/transactions/ for more info and examples.
+   */
+  suspend fun executeTransaction(action: suspend (database: TransactionalDatabase) -> Unit) {
+    require(supportChangeStreams) { "supportChangeStreams must be true for the executeTransaction() operation" }
```
Comment (vonox7) [inline src/main/kotlin/com/moshbit/katerbase/MongoDatabase.kt]: why? --> put that into the error message

**Review by vonox7 — CHANGES_REQUESTED** [review]
Cool new feature!

File: src/main/kotlin/com/moshbit/katerbase/MongoDatabase.kt (hunk 1 above)
Comment (functionaldude) [reply src/main/kotlin/com/moshbit/katerbase/MongoDatabase.kt:298]: yes, that would be much nicer, sadly there is a `notNull(..)` check on it, i can't pass a `null` here. 
In the future, if we always use a session for every mongo client that we create (which is recommended), we could clean this up

**Review by vonox7 — APPROVED** [review]

**Pull request merged**